pykeepass/fido2.py

```diff
@@ -0,0 +1,147 @@
+    PublicKeyCredentialDescriptor, PublicKeyCredentialRequestOptions, UserVerificationRequirement
+
+log = logging.getLogger(__name__)
+
+try:
+    from fido2.pcsc import CtapPcscDevice
+except ImportError:
+    CtapPcscDevice = None
+
+FIDO2_FACTOR_RPID = "fido2.keepass.nodomain"
+
+
+class NonInteractive(UserInteraction):
+
+    def __init__(self, fixed_pin):
+        self.fixed_pin = fixed_pin
+
+    def request_pin(self, permissions, rp_id):
+        return self.fixed_pin
+
+
+def _get_all_authenticators():
+    for dev in CtapHidDevice.list_devices():
+        yield dev
+    if CtapPcscDevice:
+        for dev in CtapPcscDevice.list_devices():
+            yield dev
+
+
+def _get_suitable_clients(pin_data):
+    for authenticator in _get_all_authenticators():
+        authenticator_path_string = repr(authenticator)
+
+        if isinstance(pin_data, str):
+            pin_to_use = pin_data
+        else:
+            pin_to_use = pin_data.get(authenticator_path_string, pin_data.get("*", None))
+
+        client = Fido2Client(
+            authenticator,
+            "https://{}".format(FIDO2_FACTOR_RPID),
+            user_interaction=NonInteractive(pin_to_use),
+            extension_types=[
+                HmacSecretExtension,
+                CredProtectExtension
+            ]
+        )
+
+        if "hmac-secret" in client.info.extensions and "credProtect" in client.info.extensions:
+            yield client
+
+
+class FIDOException(Exception):
+    pass
+
+
+def fido2_enroll(pin_data, already_enrolled_credentials):
+    log.info("Enrolling new FIDO2 authenticator")
+
+    # We don't care about the user ID
+    # So long as it doesn't collide with another one for the same authenticator, it's all good
+    user_id = random.randbytes(16)
+
+    chosen_client = next(_get_suitable_clients(pin_data), None)
+    if chosen_client is None:
+        raise FIDOException("Could not find an authenticator supporting the hmac-secret and credProtect extensions")
+
+    credential = chosen_client.make_credential(PublicKeyCredentialCreationOptions(
+        rp=PublicKeyCredentialRpEntity(
+            name="pykeepass",
+            id=FIDO2_FACTOR_RPID
+        ),
+        user=PublicKeyCredentialUserEntity(
+            name="keepass",
+            id=user_id,
+            display_name="KeePass"
+        ),
+        challenge=random.randbytes(32),
+        pub_key_cred_params=[
+            PublicKeyCredentialParameters(
+                type=PublicKeyCredentialType.PUBLIC_KEY,
+                alg=ES256.ALGORITHM
+            )
+        ],
+        exclude_credentials=[
+            PublicKeyCredentialDescriptor(
+                type=PublicKeyCredentialType.PUBLIC_KEY,
+                id=credential_id
+            ) for credential_id in already_enrolled_credentials
+        ],
+        extensions={
+            "hmacCreateSecret": True,
+            "credentialProtectionPolicy": CredProtectExtension.POLICY.REQUIRED,
```

Comment:
This forces the user to provide "verification" (a PIN, a fingerprint, etc - not just a touch) every time the credential is used.

This is important because the hmac-secret extension says that the keys used are silently different for UV (user verification) vs non-UV scenarios. It would not be great if the key were created with non-UV, and then later the authenticator is set to always-UV mode and so it becomes impossible to get the original key again.